File: algorithms/648. Replace Words/solution.py
class Solution(object):
    def replaceWords(self, dict, sentence):
        """
        :type dict: List[str]
        :type sentence: str
        :rtype: str
        """
        t = {}
        for word in dict:
            cur = t
            for c in word:
                cur = cur.setdefault(c, {})
            cur['$'] = word
        senwords = sentence.split()
        for i in xrange(len(senwords)):
            word = senwords[i]
            cur = t
            for c in word:
                cur = cur.get(c)
                if not cur:
                    break
                if '$' in cur:
                    senwords[i] = cur['$']
                    break
        return ' '.join(senwords)

# 124 / 124 test cases passed.
# Status: Accepted
# Runtime: 128 ms

File: algorithms/648. Replace Words/solution.cc
class Solution {
public:
    string replaceWords(vector<string>& dict, string sentence) {
        struct sTrie {
            sTrie * next[26] = {};
            string word;
            ~sTrie() {
                for ( int i = 0; i < 26; ++i )
                    if ( next[i] )
                        delete next[i];
            }
        };
        sTrie * t = new sTrie();
        for ( string & w : dict ) {
            sTrie * cur = t;
            for ( char c : w ) {
                if ( nullptr == cur->next[c - 'a'] )
                    cur->next[c - 'a'] = new sTrie();
                cur = cur->next[c - 'a'];
            }
            cur->word = w;
        }
        string result, word;
        istringstream iss(sentence);
        while ( iss >> word ) {
            sTrie * cur = t;
            for ( char c : word ) {
                if ( cur->word.size() > 0 ) {
                    word = cur->word;
                    break;
                }
                if ( nullptr == cur->next[c - 'a'] )
                    break;
                cur = cur->next[c - 'a'];
            }
            result += word + ' ';
        }
        delete t;
        result.pop_back();
        return result;
    }
};

// 124 / 124 test cases passed.
// Status: Accepted
// Runtime: 65 ms
